In [0]:
import pandas as pd
import numpy as np
import matplotlib as mp
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

In [0]:
from pyspark.sql import SparkSession
from pyspark.sql import functions as F

In [0]:
spark = SparkSession.builder.getOrCreate()

EXPLORATORY DATA ANALYSIS

Reading the Data to make it EDA ready

In [0]:
train = spark.read.options(infer_schema = True, header = True).csv("/FileStore/tables/train.csv")
test = spark.read.options(infer_schema = True, header = True).csv("/FileStore/tables/test.csv")
features = spark.read.options(infer_schema = True, header = True).csv("/FileStore/tables/features.csv")
stores = spark.read.options(infer_schema = True, header = True).csv("/FileStore/tables/stores.csv")

In [0]:
train.tail(5)

Out[5]: [Row(Store='45', Dept='98', Date='2012-09-28', Weekly_Sales='508.37', IsHoliday='FALSE'),
 Row(Store='45', Dept='98', Date='2012-10-05', Weekly_Sales='628.1', IsHoliday='FALSE'),
 Row(Store='45', Dept='98', Date='2012-10-12', Weekly_Sales='1061.02', IsHoliday='FALSE'),
 Row(Store='45', Dept='98', Date='2012-10-19', Weekly_Sales='760.01', IsHoliday='FALSE'),
 Row(Store='45', Dept='98', Date='2012-10-26', Weekly_Sales='1076.8', IsHoliday='FALSE')]

In [0]:
test.show(5)

+-----+----+----------+---------+
|Store|Dept|      Date|IsHoliday|
+-----+----+----------+---------+
|    1|   1|2012-11-02|    FALSE|
|    1|   1|2012-11-09|    FALSE|
|    1|   1|2012-11-16|    FALSE|
|    1|   1|2012-11-23|     TRUE|
|    1|   1|2012-11-30|    FALSE|
+-----+----+----------+---------+
only showing top 5 rows



Train data is from 05-02-2010 to 25-10-2012                 
Test data is from 02-11-2012 to 26-07-2013

In [0]:
train.count()

Out[7]: 421570

In [0]:
test.count()

Out[8]: 115064

In [0]:
print(f" Total number of stores: {train.select('Store').distinct().count()} \n Total number of departments: {train.select('Dept').distinct().count()}")

 Total number of stores: 45 
 Total number of departments: 81


In [0]:
features.show(5)

+-----+----------+-----------+----------+---------+---------+---------+---------+---------+-----------+------------+---------+
|Store|      Date|Temperature|Fuel_Price|MarkDown1|MarkDown2|MarkDown3|MarkDown4|MarkDown5|        CPI|Unemployment|IsHoliday|
+-----+----------+-----------+----------+---------+---------+---------+---------+---------+-----------+------------+---------+
|    1|2010-02-05|      42.31|     2.572|       NA|       NA|       NA|       NA|       NA|211.0963582|       8.106|    FALSE|
|    1|2010-02-12|      38.51|     2.548|       NA|       NA|       NA|       NA|       NA|211.2421698|       8.106|     TRUE|
|    1|2010-02-19|      39.93|     2.514|       NA|       NA|       NA|       NA|       NA|211.2891429|       8.106|    FALSE|
|    1|2010-02-26|      46.63|     2.561|       NA|       NA|       NA|       NA|       NA|211.3196429|       8.106|    FALSE|
|    1|2010-03-05|       46.5|     2.625|       NA|       NA|       NA|       NA|       NA|211.3501429|       8

In [0]:
stores.show(5)

+-----+----+------+
|Store|Type|  Size|
+-----+----+------+
|    1|   A|151315|
|    2|   A|202307|
|    3|   B| 37392|
|    4|   A|205863|
|    5|   B| 34875|
+-----+----+------+
only showing top 5 rows



Combining training and testing data so that any all transformations are done at once. Later we separate it for model feature selection and model training

In [0]:
train = train.withColumn("is_test", F.lit(0))
test = test.withColumn("Weekly_Sales", F.lit(None)).withColumn("is_test", F.lit(1))
weekly_sales = train.union(test)

In [0]:
print(f"No of rows in training data {train.count()}\nNo of rows in testing data {test.count()}\nNo of rows after merging {weekly_sales.count()}")

No of rows in training data 421570
No of rows in testing data 115064
No of rows after merging 536634


In [0]:
weekly_sales.printSchema()

root
 |-- Store: string (nullable = true)
 |-- Dept: string (nullable = true)
 |-- Date: string (nullable = true)
 |-- Weekly_Sales: string (nullable = true)
 |-- IsHoliday: string (nullable = true)
 |-- is_test: integer (nullable = false)



In [0]:
from pyspark.sql.types import IntegerType, DateType, FloatType, StringType

Joining Weekly_Sales dataframe with the features dataframe              
Note: We won't use store data as anyways each store and dept will be trained separately and thus this data won't have any effect as a feature while training the time series model

In [0]:
df_store = weekly_sales.join(features, ["Date", "Store", "IsHoliday"], how= "leftOuter")
df_store.show(5)

+----------+-----+---------+----+------------+-------+-----------+----------+---------+---------+---------+---------+---------+-----------+------------+
|      Date|Store|IsHoliday|Dept|Weekly_Sales|is_test|Temperature|Fuel_Price|MarkDown1|MarkDown2|MarkDown3|MarkDown4|MarkDown5|        CPI|Unemployment|
+----------+-----+---------+----+------------+-------+-----------+----------+---------+---------+---------+---------+---------+-----------+------------+
|2010-02-05|    1|    FALSE|   1|     24924.5|      0|      42.31|     2.572|       NA|       NA|       NA|       NA|       NA|211.0963582|       8.106|
|2010-02-12|    1|     TRUE|   1|    46039.49|      0|      38.51|     2.548|       NA|       NA|       NA|       NA|       NA|211.2421698|       8.106|
|2010-02-19|    1|    FALSE|   1|    41595.55|      0|      39.93|     2.514|       NA|       NA|       NA|       NA|       NA|211.2891429|       8.106|
|2010-02-26|    1|    FALSE|   1|    19403.54|      0|      46.63|     2.561|     

Transforming the data into appropriate datatypes

In [0]:
df_store.printSchema()

root
 |-- Date: string (nullable = true)
 |-- Store: string (nullable = true)
 |-- IsHoliday: string (nullable = true)
 |-- Dept: string (nullable = true)
 |-- Weekly_Sales: string (nullable = true)
 |-- is_test: integer (nullable = false)
 |-- Temperature: string (nullable = true)
 |-- Fuel_Price: string (nullable = true)
 |-- MarkDown1: string (nullable = true)
 |-- MarkDown2: string (nullable = true)
 |-- MarkDown3: string (nullable = true)
 |-- MarkDown4: string (nullable = true)
 |-- MarkDown5: string (nullable = true)
 |-- CPI: string (nullable = true)
 |-- Unemployment: string (nullable = true)



In [0]:
df_store = df_store.withColumn("Temperature", F.col("Temperature").cast(FloatType())).withColumn("Fuel_Price", F.col("Fuel_Price").cast(FloatType())).withColumn("MarkDown1", F.col("MarkDown1").cast(FloatType())).withColumn("MarkDown2", F.col("MarkDown2").cast(FloatType())).withColumn("MarkDown3", F.col("MarkDown3").cast(FloatType())).withColumn("MarkDown4", F.col("MarkDown4").cast(FloatType())).withColumn("MarkDown5", F.col("MarkDown5").cast(FloatType())).withColumn("CPI", F.col("CPI").cast(FloatType())).withColumn("Unemployment", F.col("Unemployment").cast(FloatType())).withColumn("Date", F.col("Date").cast(DateType())).withColumn("Weekly_Sales", F.col("Weekly_Sales").cast(FloatType())).withColumn("Store", F.col("Store").cast(IntegerType()))

DATA DICTIONARY

Store - the store number                 
Dept - the department number                 
Date - the week                           
Weekly_Sales -  sales for the given department in the given store                       
IsHoliday - whether the week is a special holiday week                            
Temperature - average temperature in the region                      
Fuel_Price - cost of fuel in the region                          
MarkDown1-5 - anonymized data related to promotional markdowns that Walmart is running. MarkDown data is only available after Nov 2011, and is not available for all stores all the time. Any missing value is marked with an NA.                 
CPI - the consumer price index                                  
Unemployment - the unemployment rate

FEATURE ENGINEERING AND FEATURE SELECTION

Ordinal Label Encoding FOR IsHoliday column

In [0]:
df_store = df_store.withColumn("IsHoliday",F.when(F.col("IsHoliday") == "True", 1).otherwise(0))

In [0]:
print("The total number of weeks without Walmart markdown data: {}".format(df_store.filter("Date <= date'2011-11-04'").select("Date").distinct().count()))

The total number of weeks without Walmart markdown data: 92


In [0]:
print("The total number of weeks in training data: {}".format(df_store.filter(F.col("is_test") == 0).select("Date").distinct().count()))

The total number of weeks in training data: 143


INSIGHT: Of 143 weeks of training data, promotional markdown data is known only for 51 weeks of data. Initially we will exclude these markdowns

In [0]:
df = df_store.drop(*("MarkDown1", "MarkDown2", "MarkDown3", "MarkDown4", "MarkDown5"))

In [0]:
df.show()

+----------+-----+---------+----+------------+-------+-----------+----------+---------+------------+
|      Date|Store|IsHoliday|Dept|Weekly_Sales|is_test|Temperature|Fuel_Price|      CPI|Unemployment|
+----------+-----+---------+----+------------+-------+-----------+----------+---------+------------+
|2010-02-05|    1|        0|   1|     24924.5|      0|      42.31|     2.572|211.09636|       8.106|
|2010-02-12|    1|        0|   1|    46039.49|      0|      38.51|     2.548|211.24217|       8.106|
|2010-02-19|    1|        0|   1|    41595.55|      0|      39.93|     2.514|211.28914|       8.106|
|2010-02-26|    1|        0|   1|    19403.54|      0|      46.63|     2.561|211.31964|       8.106|
|2010-03-05|    1|        0|   1|     21827.9|      0|       46.5|     2.625|211.35014|       8.106|
|2010-03-12|    1|        0|   1|    21043.39|      0|      57.79|     2.667|211.38065|       8.106|
|2010-03-19|    1|        0|   1|    22136.64|      0|      54.58|      2.72|211.21564|    

Handling Null Values if any

In [0]:
df.select([F.count(F.when(F.col(c).isNull(), c)).alias(c) for c in df.columns]).show()

+----+-----+---------+----+------------+-------+-----------+----------+------+------------+
|Date|Store|IsHoliday|Dept|Weekly_Sales|is_test|Temperature|Fuel_Price|   CPI|Unemployment|
+----+-----+---------+----+------------+-------+-----------+----------+------+------------+
|   0|    0|        0|   0|      115064|      0|     115064|    115064|115064|      115064|
+----+-----+---------+----+------------+-------+-----------+----------+------+------------+



Note: We see that features such as Temperature, Fuel_Price, CPI, Unemployment are not given in the test data. We'll use time series to predict these values in the future

In [0]:
df_feature = df.select(["Date", "Store", "Temperature", "Fuel_Price", "CPI", "Unemployment",  "is_test"]).distinct().sort(F.col("Date"), F.col("Store"))

In [0]:
df_feature.count()

Out[26]: 8190

In [0]:
df_feature.filter(F.col("is_test") == 1).count()

Out[27]: 1755

In [0]:
from pyspark.sql.types import StructType, StructField
schema = StructType(
    [
        StructField("Store", IntegerType()),
        StructField("Date", DateType()),
        StructField("Temperature", FloatType()),
        StructField("Fuel_Price", FloatType()),
        StructField("CPI", FloatType()),
        StructField("Unemployment", FloatType())
        
    ]
    )

In [0]:
df.select(F.dayofweek("Date")).collect()[0]

Out[29]: Row(dayofweek(Date)=6)

In [0]:
import fbprophet
from fbprophet import Prophet

In [0]:
from pyspark.sql.functions import pandas_udf, PandasUDFType
@pandas_udf(schema, PandasUDFType.GROUPED_MAP)
def forecast_feature(data):

    test = data[data["is_test"] == 1][["Date"]]
    test.columns = ["ds"]
    test["ds"] = pd.to_datetime(test["ds"])
    results_combined = pd.DataFrame(test)
    
    features = ["Temperature", "Fuel_Price", "CPI", "Unemployment"]
    
    
    
    for feature in features:
        train = data[data["is_test"] == 0][["Date",feature]]
        train.columns = ["ds", "y"]
        train["ds"] = pd.to_datetime(train["ds"])
        
        # instantiate the model
        model = Prophet()
        
        # fit the model
        model.fit(train)
        
        future_pd = model.make_future_dataframe(periods=test.shape[0],freq='W-FRI',include_history=False)
        
        # make predictions
        
        results_pd = model.predict(future_pd)
        results_pd.rename(columns = {"yhat": feature}, inplace = True)
        results_combined = pd.merge(results_combined, results_pd[["ds", feature]], on = "ds", how = "inner")
    
    # return predictions
    return pd.DataFrame({'Store': data._get_value(0, "Store"),'Date': results_combined["ds"], 'Temperature': results_combined["Temperature"], 'Fuel_Price': results_combined["Fuel_Price"],'CPI': results_combined["CPI"], 'Unemployment': results_combined["Unemployment"]})

In [0]:
results = (
    df_feature
    .groupBy('Store')
    .apply(forecast_feature)
    )

In [0]:
results = results.cache()

In [0]:
results.show()

+-----+----------+-----------+----------+---------+------------+
|Store|      Date|Temperature|Fuel_Price|      CPI|Unemployment|
+-----+----------+-----------+----------+---------+------------+
|   31|2012-11-02|  62.463303| 3.4760756|223.08025|   6.2060027|
|   31|2012-11-09|  60.323032| 3.4589472|223.04842|    6.201559|
|   31|2012-11-16|   57.43928| 3.4435263|223.01009|   6.1673317|
|   31|2012-11-23|  53.883717| 3.4265876|222.98979|   6.1267176|
|   31|2012-11-30|   50.46428| 3.4092705|223.01224|    6.109187|
|   31|2012-12-07|  48.049553|  3.401732|223.07225|    6.118647|
|   31|2012-12-14|   46.74868| 3.4127457|223.13518|    6.125961|
|   31|2012-12-21|   45.92714| 3.4378996|223.17876|   6.0930696|
|   31|2012-12-28|   45.03206|  3.463723|223.22963|    6.006984|
|   31|2013-01-04|  44.174603|  3.484071|223.34447|      5.8931|
|   31|2013-01-11|  43.746548| 3.5075586|223.54837|    5.796365|
|   31|2013-01-18|  43.662586| 3.5453598|223.79878|   5.7474074|
|   31|2013-01-25|    43.

JOB HALF DONE! We have successfully found the missing values of the features for the test data, which will now help us make the final model more comprehensive and complex as we can count on these features

Dividing the actual dataframe into test and train data, as we want to join the forecasted features into the actual data

In [0]:
test = df.filter(F.col("is_test") == 1)
train = df.filter(F.col("is_test") == 0)

In [0]:
test = test.alias("t").join(results.alias("r"), on = ["Store", "Date"], how = "leftOuter").select(F.col("t.Date"), F.col("t.Store"), F.col("t.IsHoliday"),  F.col("t.Dept"), F.col("t.Weekly_Sales"), F.col("t.is_test"), F.col("r.Temperature"), F.col("r.Fuel_Price"), F.col("r.CPI"), F.col("r.Unemployment"))

In [0]:
test.count()

Out[37]: 115064

In [0]:
df_final = train.union(test)

In [0]:
df_final = df_final.cache()

In [0]:
df_final.show(10)

+----------+-----+---------+----+------------+-------+-----------+----------+---------+------------+
|      Date|Store|IsHoliday|Dept|Weekly_Sales|is_test|Temperature|Fuel_Price|      CPI|Unemployment|
+----------+-----+---------+----+------------+-------+-----------+----------+---------+------------+
|2010-02-05|    1|        0|   1|     24924.5|      0|      42.31|     2.572|211.09636|       8.106|
|2010-02-12|    1|        0|   1|    46039.49|      0|      38.51|     2.548|211.24217|       8.106|
|2010-02-19|    1|        0|   1|    41595.55|      0|      39.93|     2.514|211.28914|       8.106|
|2010-02-26|    1|        0|   1|    19403.54|      0|      46.63|     2.561|211.31964|       8.106|
|2010-03-05|    1|        0|   1|     21827.9|      0|       46.5|     2.625|211.35014|       8.106|
|2010-03-12|    1|        0|   1|    21043.39|      0|      57.79|     2.667|211.38065|       8.106|
|2010-03-19|    1|        0|   1|    22136.64|      0|      54.58|      2.72|211.21564|    

In [0]:
df_final.select([F.count(F.when(F.col(c).isNull(), c)).alias(c) for c in df_final.columns]).show()

+----+-----+---------+----+------------+-------+-----------+----------+---+------------+
|Date|Store|IsHoliday|Dept|Weekly_Sales|is_test|Temperature|Fuel_Price|CPI|Unemployment|
+----+-----+---------+----+------------+-------+-----------+----------+---+------------+
|   0|    0|        0|   0|      115064|      0|          0|         0|  0|           0|
+----+-----+---------+----+------------+-------+-----------+----------+---+------------+



Our dataframe is ready for model training!

MODEL TRAINING.             
We will leverage the power of pandas udf to train multiple models for each store and dept parallelly

In [0]:
result_schema = StructType(
    [
        StructField("Store", IntegerType()),
        StructField("Date", DateType()),
        StructField("Dept", IntegerType()),
        StructField("Weekly_Sales", FloatType())
    ]
    )

In [0]:
from pyspark.sql.functions import pandas_udf, PandasUDFType

In [0]:
from fbprophet import Prophet

In [0]:
@pandas_udf(result_schema, PandasUDFType.GROUPED_MAP)
def sales_forecast_fbProphet(df1):
    import datetime
    df1.rename(columns = {"Date": "ds", "Weekly_Sales": "y"}, inplace = True)
    df1["ds"] = pd.to_datetime(df1["ds"])
    df_test = df1[df1["is_test"] == 1]
    df_train = df1[df1["is_test"] == 0]
    results_combined = pd.DataFrame(df_test)
    results_combined[["Store", "Dept"]] = results_combined[["Store", "Dept"]].apply(pd.to_numeric)
    
    if df_test.shape[0] > 0:
        if df_train.shape[0] >= 2:

            # Multi-variate forecasting using fbProphet

            # instantiate the model
            model = Prophet()

            #Adding regressors
            model.add_regressor("IsHoliday")
            model.add_regressor("Temperature")
            model.add_regressor("Fuel_Price")
            model.add_regressor("CPI")
            model.add_regressor("Unemployment")

            # fit the model
            model.fit(df_train[["ds", "y", "IsHoliday", "Temperature", "Fuel_Price" , "CPI", "Unemployment"]])

            future_pd = model.make_future_dataframe(periods=df_test.shape[0],freq='W-FRI',include_history=False)
            future_pd = df_test[["ds", "IsHoliday", "Temperature", "Fuel_Price" , "CPI", "Unemployment"]]

            # make predictions
            results_pd = model.predict(future_pd)
            results_combined = pd.merge(results_combined, results_pd[["ds", "yhat"]], on = "ds", how = "left")
            
            return pd.DataFrame({'Store': results_combined._get_value(0, "Store"),'Date': results_combined["ds"], 'Dept': results_combined["Dept"],'Weekly_Sales': results_combined["yhat"]})
        else:
            return pd.DataFrame({'Store': results_combined["Store"],'Date': results_combined["ds"], 'Dept': results_combined["Dept"],'Weekly_Sales': results_combined["y"]})
    else:
        date = pd.Series([])
        date = pd.to_datetime(date)
        return pd.DataFrame({'Store': pd.Series([], dtype = "int64"),'Date': date,'Dept': pd.Series([], dtype = "int64"), 'Weekly_Sales': pd.Series([], dtype = "float")})

In [0]:
final_result_fbprophet = (
    df_final
    .groupBy("Store", "Dept")
    .apply(sales_forecast_fbProphet)
)

In [0]:
final_result_fbprophet.display()

Store Date Dept Weekly_Sales 1 2012-11-02 12 9727.384 1 2012-11-09 12 9379.317 1 2012-11-16 12 9303.199 1 2012-11-23 12 9004.549 1 2012-11-30 12 8238.326 1 2012-12-07 12 7396.636 1 2012-12-14 12 7077.8086 1 2012-12-21 12 7367.0396 1 2012-12-28 12 7734.385 1 2013-01-04 12 7684.8047 1 2013-01-11 12 7350.202 1 2013-01-18 12 7292.1914 1 2013-01-25 12 7785.5073 1 2013-02-01 12 8501.563 1 2013-02-08 12 8984.11 1 2013-02-15 12 9280.93 1 2013-02-22 12 9886.493 1 2013-03-01 12 11037.826 1 2013-03-08 12 12262.327 1 2013-03-15 12 12798.223 1 2013-03-22 12 12424.764 1 2013-03-29 12 11708.096 1 2013-04-05 12 11365.495 1 2013-04-12 12 11514.174 1 2013-04-19 12 11684.429 1 2013-04-26 12 11516.93 1 2013-05-03 12 11239.013 1 2013-05-10 12 11337.499 1 2013-05-17 12 11876.853 1 2013-05-24 12 12345.141 1 2013-05-31 12 12226.692 1 2013-06-07 12 11590.2295 1 2013-06-14 12 10985.473 1 2013-06-21 12 10814.946 1 2013-06-28 12 10963.92 1 2013-07-05 12 11081.908 1 2013-07-12 12 11068.594 1 2013-07-19 12 11117.458 1 2013-07-26 12 11321.434 1 2012-11-02 23 23111.15 1 2012-11-09 23 25640.902 1 2012-11-16 23 25856.338 1 2012-11-23 23 25287.477 1 2012-11-30 23 27733.986 1 2012-12-07 23 34572.953 1 2012-12-14 23 41930.746 1 2012-12-21 23 43488.285 1 2012-12-28 23 36511.77 1 2013-01-04 23 24890.771 1 2013-01-11 23 15938.931 1 2013-01-18 23 14187.954 1 2013-01-25 23 18252.158 1 2013-02-01 23 23237.045 1 2013-02-08 23 25488.205 1 2013-02-15 23 24884.252 1 2013-02-22 23 23365.691 1 2013-03-01 23 22366.963 1 2013-03-08 23 22000.973 1 2013-03-15 23 21979.736 1 2013-03-22 23 22340.664 1 2013-03-29 23 23064.254 1 2013-04-05 23 23590.4 1 2013-04-12 23 23293.676 1 2013-04-19 23 22356.484 1 2013-04-26 23 21712.219 1 2013-05-03 23 21977.346 1 2013-05-10 23 22767.115 1 2013-05-17 23 23344.56 1 2013-05-24 23 23722.77 1 2013-05-31 23 24652.389 1 2013-06-07 23 26412.2 1 2013-06-14 23 28023.953 1 2013-06-21 23 28081.809 1 2013-06-28 23 26294.562 1 2013-07-05 23 23844.498 1 2013-07-12 23 22141.361 1 2013-07-19 23 21575.246 1 2013-07-26 23 21747.914 1 2012-11-02 27 1912.3538 1 2012-11-09 27 2006.1091 1 2012-11-16 27 1912.1329 1 2012-11-23 27 1875.4612 1 2012-11-30 27 2242.6885 1 2012-12-07 27 3044.7832 1 2012-12-14 27 3838.1956 1 2012-12-21 27 4031.448 1 2012-12-28 27 3410.0786 1 2013-01-04 27 2347.5964 1 2013-01-11 27 1499.6661 1 2013-01-18 27 1283.3956 1 2013-01-25 27 1615.8203 1 2013-02-01 27 2087.297 1 2013-02-08 27 2333.4763 1 2013-02-15 27 2258.665 1 2013-02-22 27 1990.3036 1 2013-03-01 27 1705.2302 1 2013-03-08 27 1509.7633 1 2013-03-15 27 1422.1174 1 2013-03-22 27 1402.0096 1 2013-03-29 27 1385.6785 1 2013-04-05 27 1327.5929 1 2013-04-12 27 1236.6448 1 2013-04-19 27 1166.3308 1 2013-04-26 27 1153.7711 1 2013-05-03 27 1172.6486 1 2013-05-10 27 1163.1877 1 2013-05-17 27 1108.1941 1 2013-05-24 27 1056.6906 1 2013-05-31 27 1058.0518 1 2013-06-07 27 1090.6293 1 2013-06-14 27 1085.5107 1 2013-06-21 27 1021.9514 1 2013-06-28 27 965.8061 1 2013-07-05 27 987.7596 1 2013-07-12 27 1063.1458 1 2013-07-19 27 1097.8389 1 2013-07-26 27 1065.923 1 2012-11-02 32 6787.2236 1 2012-11-09 32 7564.6084 1 2012-11-16 32 6646.14 1 2012-11-23 32 5573.8193 1 2012-11-30 32 7593.468 1 2012-12-07 32 14029.902 1 2012-12-14 32 21820.396 1 2012-12-21 32 25356.36 1 2012-12-28 32 21343.734 1 2013-01-04 32 12005.691 1 2013-01-11 32 3431.9392 1 2013-01-18 32 668.2249 1 2013-01-25 32 4123.7217 1 2013-02-01 32 10151.398 1 2013-02-08 32 14498.58 1 2013-02-15 32 15185.859 1 2013-02-22 32 12868.594 1 2013-03-01 32 9458.989 1 2013-03-08 32 6739.3057 1 2013-03-15 32 5675.4883 1 2013-03-22 32 6159.939 1 2013-03-29 32 7112.4956 1 2013-04-05 32 7262.5083 1 2013-04-12 32 6341.57 1 2013-04-19 32 5505.52 1 2013-04-26 32 6205.4834 1 2013-05-03 32 8465.175 1 2013-05-10 32 10523.737 1 2013-05-17 32 10498.992 1 2013-05-24 32 8375.748 1 2013-05-31 32 6066.781 1 2013-06-07 32 5406.769 1 2013-06-14 32 6320.5903 1 2013-06-21 32 7188.114 1 2013-06-28 32 6837.4185 1 2013-07-05 32 5758.2607 1 2013-07-12 32 5224.9985 1 2013-07-19 

In [0]:
final_result_fbprophet = final_result_fbprophet.select(F.concat_ws('_',final_result_fbprophet.Store.astype(StringType()) , final_result_fbprophet.Dept.astype(StringType()) , final_result_fbprophet.Date.astype(StringType())).alias("Id"), "Weekly_Sales").sort("Store", "Dept", "Date")

Replacing null values and negative values with zero

In [0]:
final_result_fbprophet = final_result_fbprophet.withColumn("Weekly_Sales", F.when(((F.col("Weekly_Sales") < 0) | (F.col("Weekly_Sales").isNull())), 0).otherwise(F.col("Weekly_Sales")))

In [0]:
final_result_fbprophet.cache()

Out[52]: DataFrame[Id: string, Weekly_Sales: float]

In [0]:
final_result_fbprophet.display()

Id Weekly_Sales 1_1_2012-11-02 33310.984 1_1_2012-11-09 27762.264 1_1_2012-11-16 19833.027 1_1_2012-11-23 16897.316 1_1_2012-11-30 22920.828 1_1_2012-12-07 34474.39 1_1_2012-12-14 43389.71 1_1_2012-12-21 43426.688 1_1_2012-12-28 34762.98 1_1_2013-01-04 22948.361 1_1_2013-01-11 14426.078 1_1_2013-01-18 12888.094 1_1_2013-01-25 18347.828 1_1_2013-02-01 27828.89 1_1_2013-02-08 36459.863 1_1_2013-02-15 39359.96 1_1_2013-02-22 34593.305 1_1_2013-03-01 25284.668 1_1_2013-03-08 18142.46 1_1_2013-03-15 18486.209 1_1_2013-03-22 25848.062 1_1_2013-03-29 34431.367 1_1_2013-04-05 38168.59 1_1_2013-04-12 35524.41 1_1_2013-04-19 29660.547 1_1_2013-04-26 24566.496 1_1_2013-05-03 21632.248 1_1_2013-05-10 19815.555 1_1_2013-05-17 18121.55 1_1_2013-05-24 16940.773 1_1_2013-05-31 16985.926 1_1_2013-06-07 17827.797 1_1_2013-06-14 18214.69 1_1_2013-06-21 17655.535 1_1_2013-06-28 16997.35 1_1_2013-07-05 17189.996 1_1_2013-07-12 17952.3 1_1_2013-07-19 18165.717 1_1_2013-07-26 17428.725 1_2_2012-11-02 45248.984 1_2_2012-11-09 46370.203 1_2_2012-11-16 45297.44 1_2_2012-11-23 43577.242 1_2_2012-11-30 44443.73 1_2_2012-12-07 48880.15 1_2_2012-12-14 53772.59 1_2_2012-12-21 54643.53 1_2_2012-12-28 50235.6 1_2_2013-01-04 43957.21 1_2_2013-01-11 40601.9 1_2_2013-01-18 41907.055 1_2_2013-01-25 45553.273 1_2_2013-02-01 48164.555 1_2_2013-02-08 48614.09 1_2_2013-02-15 48211.26 1_2_2013-02-22 48318.625 1_2_2013-03-01 48668.008 1_2_2013-03-08 48181.875 1_2_2013-03-15 46771.895 1_2_2013-03-22 45618.035 1_2_2013-03-29 45689.805 1_2_2013-04-05 46566.535 1_2_2013-04-12 47055.027 1_2_2013-04-19 46680.066 1_2_2013-04-26 46098.176 1_2_2013-05-03 46051.555 1_2_2013-05-10 46392.85 1_2_2013-05-17 46452.363 1_2_2013-05-24 46096.44 1_2_2013-05-31 45923.535 1_2_2013-06-07 46333.62 1_2_2013-06-14 46834.07 1_2_2013-06-21 46633.137 1_2_2013-06-28 45752.082 1_2_2013-07-05 45094.703 1_2_2013-07-12 45304.695 1_2_2013-07-19 45946.91 1_2_2013-07-26 46176.594 1_3_2012-11-02 10940.878 1_3_2012-11-09 12319.277 1_3_2012-11-16 12011.002 1_3_2012-11-23 11082.958 1_3_2012-11-30 11502.389 1_3_2012-12-07 13380.761 1_3_2012-12-14 14828.063 1_3_2012-12-21 14484.128 1_3_2012-12-28 13296.973 1_3_2013-01-04 13229.658 1_3_2013-01-11 14658.903 1_3_2013-01-18 15889.173 1_3_2013-01-25 15373.521 1_3_2013-02-01 13706.164 1_3_2013-02-08 12816.836 1_3_2013-02-15 13478.905 1_3_2013-02-22 14389.645 1_3_2013-03-01 13952.662 1_3_2013-03-08 12361.729 1_3_2013-03-15 11275.639 1_3_2013-03-22 11604.143 1_3_2013-03-29 12382.242 1_3_2013-04-05 12183.229 1_3_2013-04-12 11102.391 1_3_2013-04-19 10605.428 1_3_2013-04-26 11448.256 1_3_2013-05-03 12521.272 1_3_2013-05-10 12263.755 1_3_2013-05-17 10798.219 1_3_2013-05-24 9854.373 1_3_2013-05-31 10457.648 1_3_2013-06-07 11454.607 1_3_2013-06-14 11006.53 1_3_2013-06-21 9246.174 1_3_2013-06-28 8539.519 1_3_2013-07-05 10766.763 1_3_2013-07-12 15046.079 1_3_2013-07-19 19006.217 1_3_2013-07-26 22127.809 1_4_2012-11-02 37478.477 1_4_2012-11-09 39805.625 1_4_2012-11-16 42166.17 1_4_2012-11-23 42952.94 1_4_2012-11-30 42138.465 1_4_2012-12-07 41399.367 1_4_2012-12-14 42188.832 1_4_2012-12-21 43882.746 1_4_2012-12-28 44264.92 1_4_2013-01-04 42000.89 1_4_2013-01-11 38347.88 1_4_2013-01-18 36070.242 1_4_2013-01-25 36665.938 1_4_2013-02-01 38941.5 1_4_2013-02-08 40436.348 1_4_2013-02-15 39994.785 1_4_2013-02-22 38637.844 1_4_2013-03-01 38018.13 1_4_2013-03-08 38502.1 1_4_2013-03-15 39083.34 1_4_2013-03-22 38921.76 1_4_2013-03-29 38400.066 1_4_2013-04-05 38431.36 1_4_2013-04-12 39085.434 1_4_2013-04-19 39419.793 1_4_2013-04-26 38754.33 1_4_2013-05-03 37710.375 1_4_2013-05-10 37607.277 1_4_2013-05-17 38898.586 1_4_2013-05-24 40568.625 1_4_2013-05-31 41246.285 1_4_2013-06-07 40702.117 1_4_2013-06-14 39969.52 1_4_2013-06-21 40035.69 1_4_2013-06-28 40730.8 1_4_2013-07-05 41084.414 1_4_2013-07-12 40586.438 1_4_2013-07-19 39785.99 1_4_2013-07-26 39602.43 1_5_2012-11-02 24936.465 1_5_2012-11-09 31476.076 1_5_2012-11-16 39255.2 1_5_2012-11-23 45049.03 1_5_2012-11-30 48663.047 1_5_2012-12-07 5

In [0]:
@pandas_udf(result_schema, PandasUDFType.GROUPED_MAP)
def sales_forecast_xgboost(df2):
    
    import pandas as pd
    import xgboost
    from xgboost import XGBRegressor
    from sklearn.preprocessing import StandardScaler
    from sklearn.model_selection import GridSearchCV
    import math
    import warnings
    warnings.filterwarnings("ignore")
    
    df2["Date"] = pd.to_datetime(df2["Date"])
    df2["Weekly_Sales"].fillna(df2["Weekly_Sales"].mean(), inplace = True)

    #Adding lag as a feature
    df2["lag_1"] = df2["Weekly_Sales"].shift(1)
    df2["lag_1"].fillna("bfill", inplace = True)

    #Adding week of month as a feature
    df2['week_of_month'] = pd.to_numeric(df2["Date"].dt.day/7)
    df2['week_of_month'] = df2['week_of_month'].apply(lambda x: math.ceil(x))
    
    df2.set_index("Date", inplace = True)

    #Preparing the training data
    df_train = df2[df2["is_test"] == 0]    
    X_train = df_train[["IsHoliday", "Temperature", "Fuel_Price" , "CPI", "Unemployment"]]
    df_train.reset_index(inplace = True)
    y_train = df_train["Weekly_Sales"]

    #Preparing the test data
    df_test = df2[df2["is_test"] == 1]
    X_test = df_test[["IsHoliday", "Temperature", "Fuel_Price" , "CPI", "Unemployment"]]
    

    results_combined = pd.DataFrame(df_test)
    results_combined[["Store", "Dept"]] = results_combined[["Store", "Dept"]].apply(pd.to_numeric)

    if df_test.shape[0] > 0:
        xgb = XGBRegressor()
        xgb.fit(X_train, y_train)
        prediction = xgb.predict(X_test)
        df_test.reset_index(inplace = True)

        return pd.DataFrame({'Store': df_test["Store"],'Date': df_test["Date"], 'Dept': df_test["Dept"],'Weekly_Sales': pd.Series(prediction)})
    else:
        date = pd.Series([])
        date = pd.to_datetime(date)
        return pd.DataFrame({'Store': pd.Series([], dtype = "int64"),'Date': date,'Dept': pd.Series([], dtype = "int64"), 'Weekly_Sales': pd.Series([], dtype = "float")})

In [0]:
final_result_xgboost = (
    df_final
    .groupBy("Store", "Dept")
    .apply(sales_forecast_xgboost)
)

In [0]:
final_result_xgboost.display()

Store Date Dept Weekly_Sales 1 2012-11-02 12 9003.61 1 2012-11-09 12 8562.752 1 2012-11-16 12 9649.169 1 2012-11-23 12 8253.91 1 2012-11-30 12 7675.233 1 2012-12-07 12 8022.579 1 2012-12-14 12 7870.3887 1 2012-12-21 12 7909.701 1 2012-12-28 12 7772.1094 1 2013-01-04 12 7909.0806 1 2013-01-11 12 7673.0 1 2013-01-18 12 7417.5596 1 2013-01-25 12 7684.262 1 2013-02-01 12 7664.421 1 2013-02-08 12 7332.4653 1 2013-02-15 12 7175.0327 1 2013-02-22 12 7217.258 1 2013-03-01 12 6050.0938 1 2013-03-08 12 9859.918 1 2013-03-15 12 9975.413 1 2013-03-22 12 9903.64 1 2013-03-29 12 9874.238 1 2013-04-05 12 9917.274 1 2013-04-12 12 10314.3955 1 2013-04-19 12 10603.055 1 2013-04-26 12 10603.055 1 2013-05-03 12 10603.055 1 2013-05-10 12 10724.578 1 2013-05-17 12 10159.288 1 2013-05-24 12 10995.6045 1 2013-05-31 12 10995.6045 1 2013-06-07 12 9823.7705 1 2013-06-14 12 9091.574 1 2013-06-21 12 9139.2295 1 2013-06-28 12 8831.127 1 2013-07-05 12 9280.217 1 2013-07-12 12 9229.056 1 2013-07-19 12 8742.905 1 2013-07-26 12 9181.175 1 2012-11-02 23 20989.355 1 2012-11-09 23 20064.621 1 2012-11-16 23 17297.81 1 2012-11-23 23 19861.824 1 2012-11-30 23 19210.523 1 2012-12-07 23 25423.186 1 2012-12-14 23 25249.465 1 2012-12-21 23 25233.121 1 2012-12-28 23 25103.852 1 2013-01-04 23 25236.107 1 2013-01-11 23 25268.342 1 2013-01-18 23 25149.23 1 2013-01-25 23 25072.52 1 2013-02-01 23 25149.264 1 2013-02-08 23 24946.014 1 2013-02-15 23 24905.602 1 2013-02-22 23 18696.354 1 2013-03-01 23 48867.957 1 2013-03-08 23 16190.188 1 2013-03-15 23 17718.35 1 2013-03-22 23 17630.709 1 2013-03-29 23 18491.639 1 2013-04-05 23 18606.006 1 2013-04-12 23 17447.184 1 2013-04-19 23 17425.088 1 2013-04-26 23 17425.088 1 2013-05-03 23 17425.088 1 2013-05-10 23 16803.795 1 2013-05-17 23 18588.502 1 2013-05-24 23 17551.053 1 2013-05-31 23 17559.809 1 2013-06-07 23 17819.566 1 2013-06-14 23 19265.572 1 2013-06-21 23 18266.54 1 2013-06-28 23 18630.41 1 2013-07-05 23 18627.094 1 2013-07-12 23 18637.688 1 2013-07-19 23 18754.398 1 2013-07-26 23 19012.684 1 2012-11-02 27 1785.9452 1 2012-11-09 27 1742.353 1 2012-11-16 27 1876.2659 1 2012-11-23 27 1701.362 1 2012-11-30 27 2088.7495 1 2012-12-07 27 2499.6836 1 2012-12-14 27 2545.2366 1 2012-12-21 27 2545.4949 1 2012-12-28 27 2489.2317 1 2013-01-04 27 2504.4685 1 2013-01-11 27 2511.265 1 2013-01-18 27 2494.8562 1 2013-01-25 27 2515.8752 1 2013-02-01 27 2510.8926 1 2013-02-08 27 2504.7034 1 2013-02-15 27 2463.4666 1 2013-02-22 27 2044.1093 1 2013-03-01 27 2069.9229 1 2013-03-08 27 1641.9377 1 2013-03-15 27 1561.0828 1 2013-03-22 27 1267.7756 1 2013-03-29 27 1350.2513 1 2013-04-05 27 1301.7528 1 2013-04-12 27 1178.9801 1 2013-04-19 27 1212.9998 1 2013-04-26 27 1210.1593 1 2013-05-03 27 1212.9998 1 2013-05-10 27 1220.8634 1 2013-05-17 27 1021.8877 1 2013-05-24 27 1099.2993 1 2013-05-31 27 1100.4702 1 2013-06-07 27 1120.4753 1 2013-06-14 27 1172.5093 1 2013-06-21 27 1161.6855 1 2013-06-28 27 1156.3196 1 2013-07-05 27 1176.3489 1 2013-07-12 27 1201.0193 1 2013-07-19 27 1171.4141 1 2013-07-26 27 1207.4121 1 2012-11-02 32 5790.575 1 2012-11-09 32 6106.5527 1 2012-11-16 32 8726.611 1 2012-11-23 32 8205.051 1 2012-11-30 32 8433.944 1 2012-12-07 32 17869.227 1 2012-12-14 32 22421.273 1 2012-12-21 32 22653.516 1 2012-12-28 32 18324.314 1 2013-01-04 32 18313.096 1 2013-01-11 32 18329.27 1 2013-01-18 32 18289.52 1 2013-01-25 32 18306.959 1 2013-02-01 32 18238.402 1 2013-02-08 32 18122.934 1 2013-02-15 32 17984.56 1 2013-02-22 32 9567.298 1 2013-03-01 32 16672.39 1 2013-03-08 32 8521.479 1 2013-03-15 32 14040.857 1 2013-03-22 32 5481.8315 1 2013-03-29 32 5417.8413 1 2013-04-05 32 5487.5015 1 2013-04-12 32 5492.5894 1 2013-04-19 32 5098.1436 1 2013-04-26 32 5098.1436 1 2013-05-03 32 5098.1436 1 2013-05-10 32 5212.5073 1 2013-05-17 32 5789.2393 1 2013-05-24 32 10560.565 1 2013-05-31 32 10535.865 1 2013-06-07 32 13016.489 1 2013-06-14 32 4631.067 1 2013-06-21 32 4471.3633 1 2013-06-28 32 4379.4873 1 2013-07-05 32 4467.712 1 2013-07-12 32 4529.099 1 2013-07-19 

In [0]:
final_result_xgboost = final_result_xgboost.select(F.concat_ws('_',final_result_xgboost.Store.astype(StringType()) , final_result_xgboost.Dept.astype(StringType()) , final_result_xgboost.Date.astype(StringType())).alias("Id"), "Weekly_Sales").sort("Store", "Dept", "Date")

In [0]:
final_result_xgboost = final_result_xgboost.withColumn("Weekly_Sales", F.when(((F.col("Weekly_Sales") < 0) | (F.col("Weekly_Sales").isNull())), 0).otherwise(F.col("Weekly_Sales")))

In [0]:
final_result_xgboost.cache()

Out[71]: DataFrame[Id: string, Weekly_Sales: float]

In [0]:
final_result_xgboost.display()

Id Weekly_Sales 1_1_2012-11-02 24892.76 1_1_2012-11-09 25464.375 1_1_2012-11-16 26853.299 1_1_2012-11-23 26475.893 1_1_2012-11-30 25543.693 1_1_2012-12-07 36648.652 1_1_2012-12-14 38755.16 1_1_2012-12-21 42290.37 1_1_2012-12-28 30497.705 1_1_2013-01-04 32512.557 1_1_2013-01-11 33047.047 1_1_2013-01-18 27126.342 1_1_2013-01-25 30656.611 1_1_2013-02-01 29818.873 1_1_2013-02-08 29998.264 1_1_2013-02-15 26544.326 1_1_2013-02-22 20816.533 1_1_2013-03-01 25759.291 1_1_2013-03-08 15893.155 1_1_2013-03-15 16583.146 1_1_2013-03-22 15001.966 1_1_2013-03-29 14729.914 1_1_2013-04-05 14737.113 1_1_2013-04-12 16484.576 1_1_2013-04-19 16730.668 1_1_2013-04-26 15832.852 1_1_2013-05-03 16730.668 1_1_2013-05-10 16108.504 1_1_2013-05-17 29800.63 1_1_2013-05-24 19802.924 1_1_2013-05-31 18600.527 1_1_2013-06-07 17313.684 1_1_2013-06-14 17718.236 1_1_2013-06-21 16248.858 1_1_2013-06-28 17722.002 1_1_2013-07-05 17719.57 1_1_2013-07-12 17710.707 1_1_2013-07-19 17710.707 1_1_2013-07-26 18019.357 1_2_2012-11-02 47178.61 1_2_2012-11-09 46128.93 1_2_2012-11-16 50346.36 1_2_2012-11-23 46082.723 1_2_2012-11-30 45396.164 1_2_2012-12-07 49899.145 1_2_2012-12-14 43450.61 1_2_2012-12-21 43367.895 1_2_2012-12-28 43047.605 1_2_2013-01-04 43693.45 1_2_2013-01-11 43709.758 1_2_2013-01-18 43020.242 1_2_2013-01-25 43653.598 1_2_2013-02-01 44355.402 1_2_2013-02-08 43754.902 1_2_2013-02-15 43102.113 1_2_2013-02-22 44322.293 1_2_2013-03-01 46431.695 1_2_2013-03-08 49677.562 1_2_2013-03-15 45867.523 1_2_2013-03-22 45876.73 1_2_2013-03-29 48728.246 1_2_2013-04-05 42950.93 1_2_2013-04-12 41100.11 1_2_2013-04-19 40951.223 1_2_2013-04-26 40951.223 1_2_2013-05-03 40951.223 1_2_2013-05-10 40977.418 1_2_2013-05-17 43860.7 1_2_2013-05-24 43199.16 1_2_2013-05-31 43607.668 1_2_2013-06-07 43113.234 1_2_2013-06-14 45218.03 1_2_2013-06-21 45370.336 1_2_2013-06-28 45116.797 1_2_2013-07-05 46114.07 1_2_2013-07-12 46216.477 1_2_2013-07-19 45614.59 1_2_2013-07-26 45248.55 1_3_2012-11-02 8425.303 1_3_2012-11-09 8863.425 1_3_2012-11-16 9204.815 1_3_2012-11-23 9084.555 1_3_2012-11-30 9113.565 1_3_2012-12-07 8823.388 1_3_2012-12-14 8481.916 1_3_2012-12-21 8523.548 1_3_2012-12-28 8233.758 1_3_2013-01-04 8506.024 1_3_2013-01-11 8544.554 1_3_2013-01-18 8337.17 1_3_2013-01-25 8710.689 1_3_2013-02-01 11235.021 1_3_2013-02-08 11196.715 1_3_2013-02-15 10752.522 1_3_2013-02-22 11563.501 1_3_2013-03-01 11692.413 1_3_2013-03-08 11656.04 1_3_2013-03-15 11502.013 1_3_2013-03-22 11412.191 1_3_2013-03-29 11281.754 1_3_2013-04-05 11290.207 1_3_2013-04-12 10951.586 1_3_2013-04-19 10790.59 1_3_2013-04-26 10790.59 1_3_2013-05-03 10790.59 1_3_2013-05-10 10665.076 1_3_2013-05-17 12028.642 1_3_2013-05-24 12173.951 1_3_2013-05-31 12138.6045 1_3_2013-06-07 20518.477 1_3_2013-06-14 23420.986 1_3_2013-06-21 23500.803 1_3_2013-06-28 23461.912 1_3_2013-07-05 23443.027 1_3_2013-07-12 23443.268 1_3_2013-07-19 23427.32 1_3_2013-07-26 23509.834 1_4_2012-11-02 36848.73 1_4_2012-11-09 36733.168 1_4_2012-11-16 36951.098 1_4_2012-11-23 35466.56 1_4_2012-11-30 36740.176 1_4_2012-12-07 38010.4 1_4_2012-12-14 36220.637 1_4_2012-12-21 36153.543 1_4_2012-12-28 36169.457 1_4_2013-01-04 36707.324 1_4_2013-01-11 35685.504 1_4_2013-01-18 34283.082 1_4_2013-01-25 34463.895 1_4_2013-02-01 35877.953 1_4_2013-02-08 36118.57 1_4_2013-02-15 35344.293 1_4_2013-02-22 35774.695 1_4_2013-03-01 36693.355 1_4_2013-03-08 36716.004 1_4_2013-03-15 36241.168 1_4_2013-03-22 37506.816 1_4_2013-03-29 37777.77 1_4_2013-04-05 36775.195 1_4_2013-04-12 36741.457 1_4_2013-04-19 36664.676 1_4_2013-04-26 36664.676 1_4_2013-05-03 36664.676 1_4_2013-05-10 37428.223 1_4_2013-05-17 38792.492 1_4_2013-05-24 38116.92 1_4_2013-05-31 38192.156 1_4_2013-06-07 37782.72 1_4_2013-06-14 37146.22 1_4_2013-06-21 37439.11 1_4_2013-06-28 37479.633 1_4_2013-07-05 37934.766 1_4_2013-07-12 38530.92 1_4_2013-07-19 37432.023 1_4_2013-07-26 38477.03 1_5_2012-11-02 22766.959 1_5_2012-11-09 21385.889 1_5_2012-11-16 21976.287 1_5_2012-11-23 20277.902 1_5_2012-11-30 24416.072 1_5_2012-